File: problems/evaluate_division/evaluate_division_test.py
"""Test cases for the evaluate_division function."""

import pytest

from evaluate_division import evaluateDivision


@pytest.mark.parametrize(
    "equations, values, queries, expected",
    [
        (
            [["a", "b"], ["b", "c"]],
            [2.0, 3.0],
            [["a", "c"], ["b", "a"], ["a", "e"], ["a", "a"], ["x", "x"]],
            [6.00000, 0.50000, -1.00000, 1.00000, -1.00000],
        ),  # Example 1
        (
            [["a", "b"], ["b", "c"], ["bc", "cd"]],
            [1.5, 2.5, 5.0],
            [["a", "c"], ["c", "b"], ["bc", "cd"], ["cd", "bc"]],
            [3.75000, 0.40000, 5.00000, 0.20000],
        ),  # Example 2
        (
            [["a", "b"]],
            [0.5],
            [["a", "b"], ["b", "a"], ["a", "c"], ["x", "y"]],
            [0.50000, 2.00000, -1.00000, -1.00000],
        ),  # Example 3
    ],
    ids=["example_1", "example_2", "example_3"],
)
def test_evaluate_division(equations, values, queries, expected):
    assert evaluateDivision(equations, values, queries) == expected

File: problems/evaluate_division/evaluate_division_test.cc
#include "evaluate_division.h"

#include <gtest/gtest.h>
#include <string>
#include <vector>

struct EvaluateDivisionCase {
    std::string test_name;
    std::vector<std::vector<std::string>> equations;
    std::vector<double> values;
    std::vector<std::vector<std::string>> queries;
    std::vector<double> expected;
};

using EvaluateDivisionTest = ::testing::TestWithParam<EvaluateDivisionCase>;

TEST_P(EvaluateDivisionTest, TestCases) {
    const EvaluateDivisionCase &testCase = GetParam();
    const auto result = evaluateDivision(testCase.equations, testCase.values, testCase.queries);
    EXPECT_EQ(result, testCase.expected);
}

INSTANTIATE_TEST_SUITE_P(
    EvaluateDivisionTestCases, EvaluateDivisionTest,
    ::testing::Values(
        EvaluateDivisionCase{
            .test_name = "example1",
            .equations = {{"a", "b"}, {"b", "c"}},
            .values = {2.0, 3.0},
            .queries = {{"a", "c"}, {"b", "a"}, {"a", "e"}, {"a", "a"}, {"x", "x"}},
            .expected = {6.00000, 0.50000, -1.00000, 1.00000, -1.00000}},
        EvaluateDivisionCase{.test_name = "example2",
                             .equations = {{"a", "b"}, {"b", "c"}, {"bc", "cd"}},
                             .values = {1.5, 2.5, 5.0},
                             .queries = {{"a", "c"}, {"c", "b"}, {"bc", "cd"}, {"cd", "bc"}},
                             .expected = {3.75000, 0.40000, 5.00000, 0.20000}},
        EvaluateDivisionCase{.test_name = "example3",
                             .equations = {{"a", "b"}},
                             .values = {0.5},
                             .queries = {{"a", "b"}, {"b", "a"}, {"a", "c"}, {"x", "y"}},
                             .expected = {0.50000, 2.00000, -1.00000, -1.00000}}),
    [](const testing::TestParamInfo<EvaluateDivisionCase> &info) { return info.param.test_name; });
